### DISCUSSION TITLE
[Added] Feature Request: Sega Saturn Fenrir Webserver

### DISCUSSION THREAD

**Discussion by Dirge70** [retronas/discussions#44] [Ideas]
Though it's still very much a WIP, is adding support for the Sega Saturn Fenrir ODE Web server possible?
Here is a link to their Wiki:
https://github.com/Fenrir-ode/webserver

Does it look like something Retronas can implement?

**Comment by sairuk** [comment]
He thanks, potentially yeh its a bit light on build info, but we'll take a look. I've added it to the Ideas page

**Reply by Dirge70** [reply]
Thanks. It's super early days for it, but if it's on your radar once it gets fleshed out, would be lovely to have.

**Comment by sairuk** [comment]
This has been added to the testing branch, we need some help with documentation for this so if anyone has a device and can write up a guide we can look at rolling it into the main doco

**Comment by Dirge70** [comment]
Are there logs available for the fenrir server? I'm having some issues and the developer of fenrir is asking if there are any logs to troubleshoot

**Reply by sairuk** [reply]
we run this as a systemd service, so we have the journals which are fairly basic. Unfortunately the application will crash with the verbose flag enabled in this scenario.

some options for getting some info for the developer, in the terminal

* check the status of the service: ` systemctl status fenrir-ode-webserver `

* systemd status history: ` journalctl -u fenrir-ode-webserver `

* stop the service and run the application manually with `--verbose` (as your retronas user)
```
$ sudo systemctl stop fenrir-ode-webserver
$ /usr/local/bin/FenrirServer --dir /data/retronas/fenrir-ode-webserver --port 31994 --verbose
```

unfortunately neither of us have one of these devices to test with so the best we can offer at the moment is to ensure the application was listening and enumerates content

```
$ /usr/local/bin/FenrirServer --dir /data/retronas/fenrir-ode-webserver --port 31994 --verbose
23:12:11 INFO  /tmp/webserver/src/server.c:41: Cache cd image folder
23:12:11 INFO  /tmp/webserver/src/menu.http.c:80: add[0]: /data/retronas/fenrir-ode-webserver/test.iso test.iso
23:12:11 INFO  /tmp/webserver/src/menu.http.c:80: add[1]: /data/retronas/fenrir-ode-webserver/test2.iso test2.iso
23:12:11 INFO  /tmp/webserver/src/server.c:45: start http server
2022-07-18 22:12:11 2 mongoose.c:3328:mg_listen  1 accepting on http://0.0.0.0:31994 (port 31994)
```

```
$ telnet retronas 31994
Trying 192.168.0.176...
Connected to retronas.
Escape character is '^]'.
```

we get "Nothing here" when accessing the endpoint through a browser, if this is intentional it may be an opportunity for the dev to use that space as a log window for end users to access output for troubleshooting.

![image](https://user-images.githubusercontent.com/526372/179626789-b1280975-fe44-41d7-9317-74cff9251358.png)

**Reply by Dirge70** [reply]
Thank you. I was able to get it to connect, but it's a little choppy and slow. It could be the wifi I have isn't up to the challenge, but I'll see if the developer is able to use this information.

**Comment by the0remora** [comment]
I got my this set up and working with my Fenrir today. It works fine, the only issues I ran into are I believe upstream or with my WiFi. If anything it may need a "Refresh Games list" "Button" possibly under tools. I just needs to restart the fenrir-ode-webserver service.

Ex: lag in Panzer Dragoon cutscenes, and the translation of Grandia (v1.1.0) would not make it to the cinematic before the title screen (I still need to test the untranslated version.)

**Reply by sairuk** [reply]
ok we might be able to do one better than that, systemd (where we are running the webserver from) has .path files to monitor for path changes and reload a service based on change, the only caveat would be monitors folders don't support changes to subdirs

can you tell me if the fenrir web server supports navigating subddirs? If so i can't do the automated one, and will add an item to refresh the games list as suggested

**Reply by the0remora** [reply]
The server indexes the folder recursively but displays everything in a flattened list.

**Reply by sairuk** [reply]
had a bit of a play with this. added a gamelist option RetroNAS which will do the reload

I looked at automating it with inotifywait ... can be done, we may go down this path later on if more services fall into this situation, we'll see. I need a good way to minimise service reloads otherwise i assume we could run into reload/gaming interrruptions

**Reply by davedoc1984** [reply]
How do you get this setup? Can someone put some instructions here if ok?

**Reply by the0remora** [reply]
@davedoc1984 its part of the testing branch.
>Users can opt in/out of these branches using the `Branch` tool available in the Update menu. It is suggested you only opt into the `testing` branch if you wish you provide feedback on new features we consider complete. Opting into a `feature_name` branch will not receive any support as the work is not considered complete.

**Reply by sairuk** [reply]
it was part of the testing branch its since been added to main in the last merge.

There needs to be some Fenrir side setup doco developed by the community because none of us have one, from the retronas side its a basic install

**Reply by the0remora** [reply]
I stand corrected. I also feel volunteered, so I will figure out how to get the Fenrir side instructions written up and shared here.

**Reply by sairuk** [reply]
that would be an incredible help, thank you very much.
